### PR TITLE
feat: change deployment type from gitops to non-gitops and vice-versa for chart store apps

### LINKED ISSUE #4006: Feature: need api for migrating helm apps deployed in git-ops to no-gitops
### 🔖 Feature description

As we are migrating Devtron applications from GitOps to non-GitOps using API calls. We need the API to do the same for Helm applications deployed with GitOps.

### 🎤 Pitch

Devtron's no-GitOps sync performance is impressive, so we want to migrate this feature to all environments.

### ✌️ Solution

NONE

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#5910

### PR DESCRIPTION
**Pull request by prakash100198**
# Description

Fixes #4006 

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/appStore/InstalledAppRestHandler.go [hunk 1]
```diff
@@ -843,3 +849,112 @@ func (handler *InstalledAppRestHandlerImpl) fetchResourceTreeWithHibernateForACD
...
+	}
+
+	token := r.Header.Get("token")
+
+	if ok := handler.enforcer.Enforce(token, casbin.ResourceHelmApp, casbin.ActionDelete, "*"); !ok {
+		common.WriteJsonResp(w, errors.New("unauthorized"), nil, http.StatusForbidden)
+		return
+	}
+
+	acdToken, err := handler.argoUserService.GetLatestDevtronArgoCdUserToken()
```
Comment (nishant-d) [inline api/appStore/InstalledAppRestHandler.go:884]: should be abstracted at ACD service layer

File: api/appStore/InstalledAppRestHandler.go [hunk 2]
```diff
@@ -843,3 +849,112 @@ func (handler *InstalledAppRestHandlerImpl) fetchResourceTreeWithHibernateForACD
...
+		handler.Logger.Errorw("validation err, TriggerChartStoreAppAfterMigration", "err", err, "payload",
+			deploymentAppTriggerRequest)
+
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	token := r.Header.Get("token")
+
+	if ok := handler.enforcer.Enforce(token, casbin.ResourceHelmApp, casbin.ActionDelete, "*"); !ok {
```
Comment (nishant-d) [inline api/appStore/InstalledAppRestHandler.go:935]: action.create

File: pkg/appStore/installedApp/service/FullMode/InstalledAppDeploymentTypeChangeService.go [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+		deleteDeploymentType = bean.ArgoCd
+	}
+	//if cluster unreachable return with error, this is done to handle the case when cluster is unreachable and
+	//delete req sent to argo cd the app deletion is stuck in deleting state
+	isClusterReachable, err := impl.isClusterReachable(request.EnvId)
+	if err != nil {
+		return response, err
+	}
+	if !isClusterReachable {
+		return response, &util.ApiError{HttpStatusCode: http.StatusNotFound, InternalMessage: "cluster unreachable", UserMessage: "cluster unreachable"}
```
Comment (nishant-d) [inline pkg/appStore/installedApp/service/FullMode/InstalledAppDeploymentTypeChangeService.go:106]: change status to unreachable

File: pkg/appStore/installedApp/service/FullMode/InstalledAppDeploymentTypeChangeService.go [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+		successfullyDeletedApps = appendToDeploymentChangeStatusList(successfullyDeletedApps, installedApp, "", bean.INITIATED)
+
+	}
+	return &bean.DeploymentAppTypeChangeResponse{
+		SuccessfulPipelines: successfullyDeletedApps,
+		FailedPipelines:     failedToDeleteApps,
+	}, nil
+}
+
+func (impl *InstalledAppDeploymentTypeChangeServiceImpl) isClusterReachable(envId int) (bool, error) {
```
Comment (nishant-d) [inline pkg/appStore/installedApp/service/FullMode/InstalledAppDeploymentTypeChangeService.go:316]: move it to env/cluster service

**Review by nishant-d — CHANGES_REQUESTED** [review]

File: api/appStore/InstalledAppRestHandler.go [hunk 5]
```diff
@@ -843,3 +849,112 @@ func (handler *InstalledAppRestHandlerImpl) fetchResourceTreeWithHibernateForACD
...
+	if userId == 0 || err != nil {
+		common.WriteJsonResp(w, err, "Unauthorized User", http.StatusUnauthorized)
+		return
+	}
+
+	decoder := json.NewDecoder(r.Body)
+	var migrateAndTriggerReq *bean.DeploymentAppTypeChangeRequest
+	err = decoder.Decode(&migrateAndTriggerReq)
+	if err != nil {
+		handler.Logger.Errorw("request err, MigrateDeploymentTypeForChartStore", "err", err, "payload", migrateAndTriggerReq)
```
[the remark this answers is not in the document]
Comment (prakash100198) [reply api/appStore/InstalledAppRestHandler.go]: done

File: api/appStore/InstalledAppRestHandler.go (hunk 2 above)
Comment (prakash100198) [reply api/appStore/InstalledAppRestHandler.go]: done

File: pkg/appStore/installedApp/service/FullMode/InstalledAppDeploymentTypeChangeService.go (hunk 3 above)
Comment (prakash100198) [reply pkg/appStore/installedApp/service/FullMode/InstalledAppDeploymentTypeChangeService.go]: done

**Review by nishant-d — APPROVED** [review]

**Pull request merged**